Refactor Android property key `CStr` construction to add tests

lopopolo opened the pull request:
Add a new FFI utils module that is always tested on all platforms.

Extract the C string bytes to a const and getter function, add more docs, add tests that the const can be safely turned into a `CStr`.

astraw reviewed (approved):
Looks good. Thanks.

Kijewski commented:
I updated the CI to make sure the test is executed, because I don't anyone of us will won't to try every commit an our actual phones. :)

lopopolo commented on src/lib.rs:
```diff
@@ -86,12 +86,10 @@ pub fn get_timezone() -> std::result::Result<String, crate::GetTimezoneError> {
     platform::get_timezone_inner()
 }
 
-#[cfg(test)]
-mod tests {
```
why did we unwrap this outside of the cfg'd `mod tests`?

Kijewski replied:
My thinking was, that someone might add tests to this `mod` that don't get executed because of the `not(android)` guard, even though the tests might be useful on android, too. I.e. that the `not(android)` part was too easy to overlook. I guess that's overthinking it.

Probably it would have been even smarter to add the guard to the actual test and let the `mod` stay. I'll fix that. :)

lopopolo commented on .github/workflows/rust.yml:
```diff
@@ -40,6 +40,38 @@ jobs:
       - run: cargo test --all-targets
       - run: cargo run --example stress-test
 
+  test-only:
```
instead of blowing up CI with even more checks, would y'all be open to moving the const and the test to a `ffi_utils.rs` mod or something that we keep platform agonstic?

Kijewski replied:
Yes, that would make sense.

lopopolo replied:
done

lopopolo commented:
@astraw @Kijewski I took another stab at this to reduce some of the complexity in CI and add more tests.

astraw commented on src/ffi_utils.rs:
```diff
@@ -0,0 +1,49 @@
+
+/// Return a [`CStr`] to access the timezone from an Android system properties
+/// environment.
+pub fn android_timezone_property_name() -> &'static CStr {
```
Potential optimization: the function signature can be made `const` if the runtime check branch inside it is removed. Since the tests are below, this seems like an easy win. That said, the compiler probably figures this out anyway. Moving the unsafe to explicitly compile-time also seems safer and could then be mentioned in the `SAFETY` docstring.

I think the function should be made `pub(crate)` to signal unambiguously this is not public API.

lopopolo replied:
The CStr function is only const on recent Rust (1.59 or something, I'm on mobile and not looking at the docs)

astraw replied:
Ah, I had missed that. You are right, we can't switch to `const` until we can use rust 1.59.

I think `pub(crate)` should be good, though.

lopopolo replied:
done

Kijewski commented:
I added `Tier-` to `Tier-3` labels to match https://doc.rust-lang.org/nightly/rustc/platform-support.html. E.g Android is a Tier 2 platform.

lopopolo commented:
Alright we have two approvals, I've addressed all of the comments, and the build is green. Since this has the `Unsafe-Proposed` label, I'll merge this on October 7.

lopopolo commented:
> I added `Tier-` to `Tier-3` labels to match https://doc.rust-lang.org/nightly/rustc/platform-support.html. E.g Android is a Tier 2 platform.

Thanks for this! I went through the full history of PRs and issues and added these tags the best I could.

astraw commented:
Great, thanks @lopopolo .

This is now ready to merge. Given its `Unsafe-Proposed` status, the 7 day timer starts now. So we should merge this on 8 October.

astraw commented:
(@lopopolo somehow I missed your earlier message about merging on 7 October. Sorry about that!)
